### ISSUE TITLE
How to reset the user on logout for multi-page site

### ISSUE THREAD

**Issue by lilsweetkara** [chatwoot/chatwoot#3918]
[DISCLAIMER: First of all, I am a terrible coder. I learned a few things from my husband, who is a software engineer, but I am generally speaking terrible at this. Sorry.)

I am trying to have the chat bubble appear only for logged in users. I have a simple (crappy) node app using EJS.
I am using app.chatwoot.com

I placed the element in the header, and basically run a reset/hide it if the user is not logged in.
You mentioned in the docs that I should run `reset()` when the user logs out. The trouble I have is that the logout page is just an URL in my server that zaps the session and redirect to `/` So, no page is actually displayed.

I put thall of the code in the header, with every page, and will call reset() if the user is not logged in (or, `setUser()` if the user IS logged in). However, it feels very wasteful (or, my husband's words, "It's the worst idea since the invention of the atomic bomb")

Also, I have to have that `alreadyReset()` variable there, since 1) I need to wait for `chatwood:ready` 2 After reset, the event `chatwoot:ready` is emitted again, and reset() is run again. And again. And again. and...

Am I doing it all wrong? (Or, is it really worse than the atomic bomb?)

````

<%if(session.loggedIn) { %>

<script>
    (function(d,t) {
      var BASE_URL="https://app.chatwoot.com";
      var g=d.createElement(t),s=d.getElementsByTagName(t)[0];
      g.src=BASE_URL+"/packs/js/sdk.js";
      g.defer = true;
      g.async = true;
      s.parentNode.insertBefore(g,s);
      g.onload=function(){
        window.chatwootSDK.run({
          websiteToken: 'XXXXXXXXXXXXXXXX,
          baseUrl: BASE_URL
        })
      }
    })(document,"script");

    window.addEventListener('chatwoot:ready', function () {
        debugger
        window.$chatwoot.setUser('<%=session.loggedInEmail%>', {
            email: '<%=session.loggedInEmail%>',
            name: '<%=session.loggedInName%> <%=session.subscriber ? 'PAID' : 'FREE'%>' 
            identifier_hash: '<%=hash%>',
        });
    });        

  </script>
  
  <% } else { // This happens if user is NOT logged in  %>    
    <script>
      let alreadyReset = false
      window.chatwootSettings = {
        showPopoutButton: false
      };      
        (function(d,t) {
          var BASE_URL="https://app.chatwoot.com";
          var g=d.createElement(t),s=d.getElementsByTagName(t)[0];
          g.src=BASE_URL+"/packs/js/sdk.js";
          g.defer = true;
          g.async = true;
          s.parentNode.insertBefore(g,s);
          g.onload=function(){
            window.chatwootSDK.run({
              websiteToken: 'XXXXXXXXXXXXXXXXX',
              baseUrl: BASE_URL
            })
          }
        })(document,"script");
    
        window.chatwootSettings = {
        showPopoutButton: false
      };
      
        window.addEventListener('chatwoot:ready', function () {
            debugger
            if (!alreadyReset){
              window.$chatwoot.reset()
              alreadyReset = true
            }

        });
      </script>
    
  <% } %>
````
Basically, if the user is NOT logged in, I still load the element, but I reset it.

**Comment by lilsweetkara** [comment]
I have the feeling I really did something that is worse than the atomic bomb, and everybody is too embarrassed to answer.
I know I am terrible at this, I won't be offended. Really.

**Comment by lilsweetkara** [comment]
The only issue I see is that this code is in the header of EVERY page. So, my users are going to call a reset() EVERY time they view a page while not logged in. I need to do it in every page, because I don't actually know, on any page load, whether the user was previously logged in or not. It just seems to be very wasteful...

**Comment by maxacarvalho** [comment]
This issue is driving me crazy 🤣 

It does not make any sense, if I call the `window.$chatwoot.reset();` the event `chatwoot:ready` keeps being fired without end, so the `window.$chatwoot.reset();` will be called again and again and again.  

It seems to me that nobody is using the `window.$chatwoot.reset();` since I find it impossible that nobody is getting the same issue, I'm simply following the instructions on the documentation and nothing else.

Here, take the snippet below, it triggers the issue:

```javascript
<script>
  (function(d,t) {
    var BASE_URL="https://my-base-url.com";
    var g=d.createElement(t),s=d.getElementsByTagName(t)[0];
    g.src=BASE_URL+"/packs/js/sdk.js";
    g.defer = true;
    g.async = true;
    s.parentNode.insertBefore(g,s);
    g.onload=function(){
      window.chatwootSDK.run({
        websiteToken: 'my-website-token',
        baseUrl: BASE_URL
      })
    }
  })(document,"script");

  window.addEventListener('chatwoot:ready', function () {
    console.log('chatwoot:ready');
    window.$chatwoot.reset();
  });
</script>
```

As you can see above, that's just copy and paste from the Chatwoot docs, nothing else, but the log message ` console.log('chatwoot:ready');`, will keep firing again and again until I remove it from the code.  

So please, help us out, how can we `reset` the session? Please provide a step-by-step explanation, or a snippet of code with an example, because so far, the only solution that I see is to:

1. Load an HTML page and run the `window.$chatwoot.reset();`
2. Redirect the user to another page where there's no `window.$chatwoot.reset();`

But as you can imagine, that's not an applicable solution for us.

**Comment by maxacarvalho** [comment]
If I understood correctly, the `reset` method happens here [https://github.com/chatwoot/chatwoot/blob/23965fbaa3c729ab8da21c015e2682f1efb16422/app/javascript/packs/sdk.js#L108](https://github.com/chatwoot/chatwoot/blob/23965fbaa3c729ab8da21c015e2682f1efb16422/app/javascript/packs/sdk.js#L108).

Every time that the `reset` is fired a new `chatwoot:ready` event is fired, and that causes an infinite loop.  
I think that what we need here is a `unsetUser` method.

**Comment by maxacarvalho** [comment]
Any update here?

**Comment by sojan-official** [comment]
@maxacarvalho Thanks for debugging. We are taking a look at this.

**Comment by lilsweetkara** [comment]
@sojan-official Sojan, I just wanted to apologise for not being able to provide a good report like maxcarvalho did. I sure did my best. Thanks for helping.

**Comment by nithindavid** [comment]
@lilsweetkara @maxacarvalho Thanks for finding this. The `window.$chatwoot.reset` and `ready` goes into loop. Will create an issue

**Comment by pranavrajs** [comment]
@maxacarvalho 

We use this method on our prod instance and as mentioned before we are calling it only when the user clicks on Logout, and not on chatwoot:ready / page load. I guess the problem here was that the reset method was not designed to be used inside `chatwoot:ready`. I think this is a product bug that should have been captured before. Thanks for reporting.

**Comment by lilsweetkara** [comment]
Did this fix make it in the latest release I saw a couple of days back? Thank you!

**Comment by pranavrajs** [comment]
@lilsweetkara #4079 should fix this behaviour.

**Comment by lilsweetkara** [comment]
Hi,

Ah! Is it already released to the cloud version of chatwoot?

Thank you!!!

Kara

**Comment by pranavrajs** [comment]
@lilsweetkara Yes, it is deployed. Closing this as this is completed via https://github.com/chatwoot/chatwoot/pull/4079
